blog-posts/cdk-scheduler-presentation/cdk-scheduler-presentation.md

```diff
@@ -0,0 +1,91 @@
+---
+published: false
+title: 'CDK Scheduler'
+cover_image: ./assets/cover-photo-never-be-late.jpg
+description: 'Description of the article'
+tags: serverless, scheduling, cdk, cdk-construct
+series:
+canonical_url:
+---
+
+# A serverless solution to just-in-time scheduling
+
+TLDR; We published a **CDK construct** to schedule events precisely.
+Documentation is available on [the project's github](https://github.com/guiyom-e/cdk-scheduler)
+
+When I worked on a serverless CMS with Guillaume, we had to find a way to schedule an event (publishing a new article)
+at a precise time with serverless tools.
+AWS doesn't offer a built-in service to do so.
+We found straight-forward options but none as precise as we wished.
+We came up with a solution of our own.
+We published a CDK construct for you to add our solution to your stack without boilerplate.
+
+## Naive approaches to serverless scheduling
+
+### Scheduling a lambda to run every minute
+
+A first approach is to use **lambda schedule trigger** feature.
+Lambdas can be triggered every down to every minute.
+Every minute, you can query your database to find if any events need to be published now.
+If so, run the action you planned (like publishing an article), else do nothing.
+
+![Trigger a lambda every minute to check if any events needs to be scheduled](./assets/naive-approach-lambda.jpg)
+
+A lambda can be triggered down to every minute and the trigger time has a **one-minute precision** guarantee.
+It's good enough for some use-cases but not if you are looking for a highly precise method.
+
+### Using DynamoDB time-to-live and streams
+
+DynamoDB comes with a time-to-live feature.
+You can add a **`ttl` attribute** to an entry.
+At said time, AWS will remove the entry form your table.
+If you have a stream plugged to your table you can play an action at scheduled time.
+
+However, this approach is **even less precise** than the first.
+AWS only guarantees a forty-eight-hour precision on the time-to-live date.
+In practice, users observe a ten-minute delay but it's not guaranteed.
+If, like us, you work on time-sensitive events, you will want to avoid this method.
+
+![Use Dynaomdb time-to-live feature to trigger events with one-day precision](./assets/naive-approach-ttl.jpg)
+
+## Introducing cdk-scheduler
+
+To match our just-in-time ambition, we leveraged the SQS delay feature.
+When publishing on an SQS queue you can set a delay that can go up to fifteen minutes.
+
+![cdk-scheduler leverages SQS delay feature to trigger events precisely](./assets/cdk-scheduler-architecture.jpg)
+
+Instead of starting a lambda every minute, we can do it every fifteen minutes.
+This method has a **one-second precision gap**!
+
+In case you're in hurry to schedule, we also added a "near-future" handler.
+This second lambda is plugged on a DynamoDB stream and handles events that are to be scheduled in less than fifteen minutes after their creation.
```

Comment:
"this second lambda" -> je trouve pas hyper clair d'ou vient cette lambda ? C'est dans l'implem de la fonctionnalité near-future ?

Reply:
et je comprends pas bien si avec le near-future handler on garde la même précision ?